numba/core/typing/new_builtins.py

```diff
@@ -127,53 +131,506 @@ def generic(self, args, kws):
+    return impl
+
+
+@overload_method(types.PythonInteger, "__add__")
+@overload_method(types.PythonInteger, "__radd__")
+def py_int__add__(self, other):
+    # evaluate type
+    def impl(self, other):
+        if isinstance(other, (bool, int)) and hasattr(other, "__int__"):
+            asint = other.__int__()
+            if asint is NotImplemented:
+                return NotImplemented
+            else:
+                return int_add_int(self, asint)
+        else:
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonFloat, "__add__")
+@overload_method(types.PythonFloat, "__radd__")
+def py_float__add__(self, other):
+    # evaluate type
+    def impl(self, other):
+        if hasattr(other, "__float__"):
+            asfloat = other.__float__()
+            if asfloat is NotImplemented:
+                return NotImplemented
+            else:
+                return float_add_float(self, asfloat)
+        else:
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonComplex, "__add__")
+@overload_method(types.PythonComplex, "__radd__")
+def py_complex__add__(self, other):
+        # evaluate type
+        def impl(self, other):
+            if hasattr(other, "__float__"):
+                as_float = other.__float__()
+                if as_float is NotImplemented:
+                    return NotImplemented
+                else:
+                    return complex_add_complex(self, complex(as_float))
+            elif hasattr(other, "__complex__"):
+                as_complex = other.__complex__()
+                if as_complex is NotImplemented:
+                    return NotImplemented
+                else:
+                    return complex_add_complex(self, as_complex)
+            else:
+                return NotImplemented
+        return impl
+
+
+FROM_DTYPE = {
+    np.dtype('bool'): types.np_bool_,
+    np.dtype('int8'): types.np_int8,
+    np.dtype('int16'): types.np_int16,
+    np.dtype('int32'): types.np_int32,
+    np.dtype('int64'): types.np_int64,
+
+    np.dtype('uint8'): types.np_uint8,
+    np.dtype('uint16'): types.np_uint16,
+    np.dtype('uint32'): types.np_uint32,
+    np.dtype('uint64'): types.np_uint64,
+
+    np.dtype('float32'): types.np_float32,
+    np.dtype('float64'): types.np_float64,
+    np.dtype('float16'): types.np_float16,
+    np.dtype('complex64'): types.np_complex64,
+    np.dtype('complex128'): types.np_complex128,
+
+    np.dtype(object): types.pyobject,
+}
+
+binop_cache = {
+    "__add__": {},
+    "__radd__": {}
+}
+
+@intrinsic
+def np_bool_add_bool(tyctx, boolxty, boolyty):
+    sig = types.np_bool_(boolxty, boolyty)
+    def codegen(cgctx, builder, sig, llargs):
+        return builder.or_(*llargs)
+
+    return sig, codegen
+
+def find_np_res_type(op, op_cache, argtys):
+    if argtys in op_cache[op]:
+        return op_cache[op][argtys]
+
+    argvals = [argty.cast_python_value(0) for argty in argtys]
```

Comment:
Need to "soften" the error for when types don't have `.cast_python_value`

I hacked something like below to make my example work:
```diff
@@ -503 +503,6 @@ def find_np_res_type(op, op_cache, argtys):
-    argvals = [argty.cast_python_value(0) for argty in argtys]
+    try:
+        argvals = [argty.cast_python_value(0) for argty in argtys]
+    except NotImplementedError:
+        e = errors.NumbaTypeError(f"cast_python_value({argtys})")
+        # print(e)
+        raise e
```